### PR TITLE
[WIP] 1.2 Test Split - Modules own tests

### PR DESCRIPTION
**Pull request by jaxwilko**
@LukeTowers told me to...

This PR splits out tests into their related modules, allowing for each module to be targeted using the `winter:test` command. Added to this is also proper namespacing and a bunch of minor bug fixes and tweaks.

### CODE REVIEW

File: modules/backend/tests/widgets/FormTest.php [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            return ['dropdown', 'options'];
+        }
+
+        public function staticMethodOptions()
+        {
+            return ['static', 'method'];
+        }
+    }
+
+
```
Comment (LukeTowers) [inline modules/backend/tests/widgets/FormTest.php]:
```suggestion
```

File: modules/cms/tests/classes/AssetTest.php [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
 use Cms\Classes\Asset;
+use Cms\Classes\Theme;
 
 class AssetTest extends TestCase
 {
     public function testLoad()
     {
         $theme = Theme::load('test');
-
+
```
Comment (LukeTowers) [inline modules/cms/tests/classes/AssetTest.php]:
```suggestion

```

File: modules/backend/tests/widgets/FormTest.php [hunk 3]
```diff
@@ -0,0 +1,242 @@
+<?php
+
+namespace Backend\Tests\Widgets
+{
+    use System\Tests\Bootstrap\PluginTestCase;
+    use Illuminate\Database\Eloquent\Model;
```
Comment (LukeTowers) [inline modules/backend/tests/widgets/FormTest.php]:
```suggestion
    use Winter\Storm\Database\Model;
```

File: modules/cms/tests/classes/ComponentManagerTest.php [hunk 4]
```diff
@@ -1,24 +1,28 @@
...
+namespace Cms\Tests\Classes;
+
+use System\Tests\Bootstrap\TestCase;
 use Cms\Classes\CodeParser;
 use Cms\Classes\ComponentManager;
+use Cms\Classes\Controller;
+use Cms\Classes\Layout;
+use Cms\Classes\Page;
+use Cms\Classes\Theme;
+use function base_path;
```
Comment (LukeTowers) [inline modules/cms/tests/classes/ComponentManagerTest.php]: is this actually required or just a PHPStorm artifact?

File: modules/cms/tests/classes/PageTest.php [hunk 5]
```diff
@@ -11,7 +14,7 @@ public function testResolveMenuItem()
         $theme = Theme::load('test');
         $controller = new Controller;
 
-        $item = (object) [
+        $item = (object)[
```
Comment (LukeTowers) [inline modules/cms/tests/classes/PageTest.php]:
```suggestion
        $item = (object) [
```

**Comment by bennothommo** [comment]
Hey guys, 2 thoughts on this -

- Are we going to be forcing everyone to have to add `use System\Tests\Bootstrap\PluginTestCase` to their plugin test cases? I think we should probably still have that available as a global alias if possible.
- With the `tests` directories within the modules, we may have to look into using a `.gitattributes` file on each module with `export-ignore` definitions so that the tests aren't included on each install through Composer. (see https://github.com/composer/composer/issues/9153 and https://github.com/laravel/framework/blob/9.x/.gitattributes#L9-L23)

File: modules/cms/tests/classes/RouterTest.php [hunk 6]
```diff
@@ -1,13 +1,18 @@
 <?php
 
+namespace Cms\Tests\Classes;
+
+use System\Tests\Bootstrap\TestCase;
 use Cms\Classes\Router;
 use Cms\Classes\Theme;
+use ReflectionClass;
+use function count;
```
Comment (LukeTowers) [inline modules/cms/tests/classes/RouterTest.php]: Another PHPStorm artifact?

File: modules/cms/tests/classes/ThemeTest.php [hunk 7]
```diff
@@ -1,10 +1,17 @@
 <?php
 
+namespace Cms\Tests\Classes;
+
+use System\Tests\Bootstrap\TestCase;
 use Cms\Classes\Theme;
+use Config;
+use Event;
+use function base_path;
+use const PHP_OS_FAMILY;
```
Comment (LukeTowers) [inline modules/cms/tests/classes/ThemeTest.php]: PHPStorm?

File: modules/system/tests/classes/ImageResizerTest.php [hunk 8]
```diff
@@ -82,7 +89,7 @@ public function testConfiguration()
         ], $imageResizer->getConfig());
 
         // Resize with an customised defaults
-        $this->markTestSkipped('@TODO: Something is going funky with serialized closures in PHP 8.0 / Laravel 9.x');
+//        $this->markTestSkipped('@TODO: Something is going funky with serialized closures in PHP 8.0 / Laravel 9.x');
```
Comment (LukeTowers) [inline modules/system/tests/classes/ImageResizerTest.php]:
```suggestion
```

File: modules/system/tests/fixtures/themes/apitest/.gitignore [hunk 9]
```diff
@@ -0,0 +1,2 @@
+*.htm
+testobjects/*.htm
```
Comment (LukeTowers) [inline modules/system/tests/fixtures/themes/apitest/.gitignore:2]:
```suggestion
testobjects/*.htm

```

**Review by LukeTowers — CHANGES_REQUESTED** [review]

**Comment by LukeTowers** [comment]
@bennothommo 

1. Yes, we are. Any existing projects will still have the original tests folder and they can keep using that if they want.
2. I want the tests folders for the modules distributed via Composer, that way the `winter:test` command can always be used and our testing suite can stay up to date with the rest of the platform. Right now if we make any changes to the testing suite or the base classes the only way users can get those changes is by manually downloading them to update their project tests folder.

**Comment by bennothommo** [comment]
> @bennothommo
> 
>     1. Yes, we are. Any existing projects will still have the original tests folder and they can keep using that if they want.

As long as that doesn't create any conflicts, all good!

>     2. I want the tests folders for the modules distributed via Composer, that way the `winter:test` command can always be used and our testing suite can stay up to date with the rest of the platform. Right now if we make any changes to the testing suite or the base classes the only way users can get those changes is by manually downloading them to update their project tests folder.

I suppose it will be handy to not have to instruct people to download or updates their test folders anymore.

File: modules/cms/tests/classes/ComponentManagerTest.php (hunk 4 above)
Comment (jaxwilko) [reply modules/cms/tests/classes/ComponentManagerTest.php]: Probably just phpstorm being helpful. Although technically I think we should be importing global functions but meh

File: modules/cms/tests/classes/RouterTest.php (hunk 6 above)
Comment (jaxwilko) [reply modules/cms/tests/classes/RouterTest.php]: As above

Comment (jaxwilko) [reply modules/cms/tests/classes/RouterTest.php]: Although I prefer to `use ReflectionClass;` over `new \ReflectionClass();`. Your call though, not massively bothered.

File: modules/cms/tests/classes/ComponentManagerTest.php (hunk 4 above)
Comment (LukeTowers) [reply modules/cms/tests/classes/ComponentManagerTest.php]: Let's not import global functions. We don't do it anywhere else and it's not technically required right now.

File: modules/cms/tests/classes/RouterTest.php (hunk 6 above)
Comment (LukeTowers) [reply modules/cms/tests/classes/RouterTest.php]: I'm fine with the importing of the global classes, just not the functions.

File: modules/cms/tests/classes/ComponentManagerTest.php (hunk 4 above)
Comment (jaxwilko) [reply modules/cms/tests/classes/ComponentManagerTest.php]: Totally fine with that, i'll strip them out and push it up this evening :)

**Comment by jaxwilko** [comment]
Removed global func/const imports in d1ff0945d6959097f453fbb736d447a6efa2d966

**Review by LukeTowers — APPROVED** [review]

**Pull request merged**